### PR TITLE
[DOXIA-714] Emit proper markdown for table alignment and figures

### PR DESCRIPTION
**Pull request by kwin**
Fix markup for inline styles for "em" and "strong"

### CODE REVIEW

File: doxia-core/src/test/java/org/apache/maven/doxia/sink/impl/AbstractSinkTest.java [hunk 1]
```diff
@@ -644,10 +645,65 @@ public void testTable() {
         if (isXmlSink()) {
             assertThat(wrapXml(actual), isIdenticalTo(wrapXml(expected)));
         } else {
-            assertEquals(actual, expected);
+            assertEquals(expected, actual);
```
Comment (michael-o) [inline doxia-core/src/test/java/org/apache/maven/doxia/sink/impl/AbstractSinkTest.java:648]: Can you move these to a separate PR? Good catch, BTW.

File: doxia-core/src/test/java/org/apache/maven/doxia/sink/impl/AbstractSinkTest.java [hunk 2]
```diff
@@ -644,10 +645,65 @@ public void testTable() {
...
+     */
+    @Test
+    public void testTableWithHeader() {
+        int[] justify = {Sink.JUSTIFY_LEFT, Sink.JUSTIFY_RIGHT, Sink.JUSTIFY_CENTER};
+        sink.table();
+        sink.tableRows(justify, false);
+        try (IntStream cellStream = getCellStreamForNewRow(3)) {
+            cellStream.forEach(col -> {
+                sink.tableHeaderCell();
+                sink.text("header" + col);
```
Comment (michael-o) [inline doxia-core/src/test/java/org/apache/maven/doxia/sink/impl/AbstractSinkTest.java:668]: `"header"` should end with an underscope, just like the rows.

File: doxia-modules/doxia-module-apt/src/main/java/org/apache/maven/doxia/module/apt/AptSink.java [hunk 3]
```diff
@@ -817,6 +820,9 @@ public void text(String text, SinkEventAttributes attributes) {
         } else {
             content(text);
         }
+        if (attributes != null) {
+            inline_();
+        }
```
Comment (michael-o) [inline doxia-modules/doxia-module-apt/src/main/java/org/apache/maven/doxia/module/apt/AptSink.java:825]: Is that a bug we haven't noticed before?

File: doxia-modules/doxia-module-xdoc/src/test/java/org/apache/maven/doxia/module/xdoc/XdocSinkTest.java [hunk 4]
```diff
@@ -208,6 +208,19 @@ protected String getTableBlock(String cell, String caption) {
                 + "</td></tr></table>";
     }
 
+    @Override
+    protected String getTableWithHeaderBlock(String... rowPrefixes) {
+        return "<table border=\"0\">\n<tr>\n<th>" + rowPrefixes[0] + "0</th>\n<th>" + rowPrefixes[0] + "1</th>\n<th>"
```
Comment (michael-o) [inline doxia-modules/doxia-module-xdoc/src/test/java/org/apache/maven/doxia/module/xdoc/XdocSinkTest.java:213]: That border attribute should already be code by now. Is it still emitted? It is deprecated and one should use CSS.

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java [hunk 5]
```diff
@@ -45,12 +48,9 @@ public class MarkdownSink extends AbstractTextSink implements MarkdownMarkup {
     // Instance fields
     // ----------------------------------------------------------------------
 
-    /**  A buffer that holds the current text when headerFlag or bufferFlag set to <code>true</code>. */
+    /**  A buffer that holds the current text when headerFlag or bufferFlag set to <code>true</code>. This value of this buffer is already escaped. */
```
Comment (michael-o) [inline doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java:51]: The content of this buffer...

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java [hunk 6]
```diff
@@ -69,24 +69,21 @@ public class MarkdownSink extends AbstractTextSink implements MarkdownMarkup {
     /**  tableCaptionFlag. */
     private boolean tableCaptionFlag;
 
-    /**  tableCellFlag. */
+    /** tableCellFlag, set to true inside table (header) cells */
```
Comment (michael-o) [inline doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java:72]: `{@code true}`

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java [hunk 7]
```diff
@@ -69,24 +69,21 @@ public class MarkdownSink extends AbstractTextSink implements MarkdownMarkup {
...
     private boolean tableCellFlag;
 
+    /** tableRowHeaderFlag, set to true for table rows containing at least one table header cell  */
```
Comment (michael-o) [inline doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java:75]: Same here

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java [hunk 8]
```diff
@@ -69,24 +69,21 @@ public class MarkdownSink extends AbstractTextSink implements MarkdownMarkup {
...
+    private boolean tableHeaderCellFlag;
+
     /**  headerFlag. */
     private boolean headerFlag;
 
-    /**  bufferFlag. */
+    /**  bufferFlag, set to true in certain elements to prevent direct writing during {@link #text(String, SinkEventAttributes)} */
```
Comment (michael-o) [inline doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java:81]: Same here

**Review by michael-o — CHANGES_REQUESTED** [review]

File: doxia-core/src/test/java/org/apache/maven/doxia/sink/impl/AbstractSinkTest.java (hunk 1 above)
Comment (kwin) [reply doxia-core/src/test/java/org/apache/maven/doxia/sink/impl/AbstractSinkTest.java:648]: No, sorry, too much effort for just fixing a glitch in a test class.

File: doxia-modules/doxia-module-apt/src/main/java/org/apache/maven/doxia/module/apt/AptSink.java (hunk 3 above)
Comment (kwin) [reply doxia-modules/doxia-module-apt/src/main/java/org/apache/maven/doxia/module/apt/AptSink.java:825]: Yes, similar to https://issues.apache.org/jira/browse/DOXIA-706

File: doxia-modules/doxia-module-xdoc/src/test/java/org/apache/maven/doxia/module/xdoc/XdocSinkTest.java (hunk 4 above)
Comment (kwin) [reply doxia-modules/doxia-module-xdoc/src/test/java/org/apache/maven/doxia/module/xdoc/XdocSinkTest.java:213]: This is just what is emitted currently for the newly added test in the `AbstractParserTest`. Fixing the output should be done separately if you consider it wrong.

File: doxia-modules/doxia-module-apt/src/main/java/org/apache/maven/doxia/module/apt/AptSink.java (hunk 3 above)
Comment (michael-o) [reply doxia-modules/doxia-module-apt/src/main/java/org/apache/maven/doxia/module/apt/AptSink.java:825]: Shouldn't this be separate then?

File: doxia-modules/doxia-module-xdoc/src/test/java/org/apache/maven/doxia/module/xdoc/XdocSinkTest.java (hunk 4 above)
Comment (michael-o) [reply doxia-modules/doxia-module-xdoc/src/test/java/org/apache/maven/doxia/module/xdoc/XdocSinkTest.java:213]: Alright, let's leave as-is.

File: doxia-modules/doxia-module-apt/src/main/java/org/apache/maven/doxia/module/apt/AptSink.java (hunk 3 above)
Comment (kwin) [reply doxia-modules/doxia-module-apt/src/main/java/org/apache/maven/doxia/module/apt/AptSink.java:825]: Ok -> https://issues.apache.org/jira/browse/DOXIA-715

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java (hunk 5 above)
Comment (kwin) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java]: Fixed.

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java (hunk 6 above)
Comment (kwin) [reply doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java]: Fixed

File: doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java [hunk 9]
```diff
@@ -45,12 +48,9 @@ public class MarkdownSink extends AbstractTextSink implements MarkdownMarkup {
     // Instance fields
     // ----------------------------------------------------------------------
 
-    /**  A buffer that holds the current text when headerFlag or bufferFlag set to <code>true</code>. */
+    /**  A buffer that holds the current text when headerFlag or bufferFlag set to <code>true</code>. This content of this buffer is already escaped. */
```
Comment (michael-o) [inline doxia-modules/doxia-module-markdown/src/main/java/org/apache/maven/doxia/module/markdown/MarkdownSink.java:51]: The content...

**Review by michael-o — APPROVED** [review]
One nit and the rest is just fine...

**Comment by jira-importer** [comment]
Resolve #850
